This is TOMOYO Linux patch for Debian Lenny.

Source code for this patch is "apt-get install linux-source-2.6.26"
---
 arch/ia64/ia32/sys_ia32.c       |    3 +++
 arch/mips/kernel/ptrace32.c     |    2 ++
 fs/compat.c                     |    2 +-
 fs/compat_ioctl.c               |    3 +++
 fs/exec.c                       |   10 +++++++++-
 fs/fcntl.c                      |    4 ++++
 fs/ioctl.c                      |    2 ++
 fs/namei.c                      |   26 ++++++++++++++++++++++++++
 fs/namespace.c                  |    8 ++++++++
 fs/open.c                       |   24 ++++++++++++++++++++++++
 fs/proc/proc_misc.c             |    1 +
 fs/stat.c                       |    2 ++
 include/linux/init_task.h       |    9 +++++++++
 include/linux/sched.h           |    6 ++++++
 include/linux/security.h        |    1 +
 kernel/compat.c                 |    2 ++
 kernel/fork.c                   |    5 +++++
 kernel/kexec.c                  |    3 +++
 kernel/module.c                 |    5 +++++
 kernel/ptrace.c                 |    4 ++++
 kernel/sched.c                  |    2 ++
 kernel/signal.c                 |    8 ++++++++
 kernel/sys.c                    |   10 ++++++++++
 kernel/sysctl.c                 |    3 +++
 kernel/time.c                   |    4 ++++
 kernel/time/ntp.c               |    3 +++
 net/ipv4/inet_connection_sock.c |    3 +++
 net/ipv4/inet_hashtables.c      |    3 +++
 net/ipv4/raw.c                  |    5 +++++
 net/ipv4/udp.c                  |    8 ++++++++
 net/ipv6/raw.c                  |    5 +++++
 net/ipv6/udp.c                  |    5 +++++
 net/socket.c                    |   18 ++++++++++++++++++
 net/unix/af_unix.c              |    6 ++++++
 security/Kconfig                |    2 ++
 security/Makefile               |    3 +++
 36 files changed, 208 insertions(+), 2 deletions(-)

--- linux-2.6.26-27.orig/arch/ia64/ia32/sys_ia32.c
+++ linux-2.6.26-27/arch/ia64/ia32/sys_ia32.c
@@ -51,6 +51,7 @@
 #include <asm/types.h>
 #include <asm/uaccess.h>
 #include <asm/unistd.h>
+#include <linux/ccsecurity.h>
 
 #include "ia32priv.h"
 
@@ -1742,6 +1743,8 @@ sys32_ptrace (int request, pid_t pid, un
 	struct task_struct *child;
 	unsigned int value, tmp;
 	long i, ret;
+	if (ccs_ptrace_permission(request, pid))
+		return -EPERM;
 
 	lock_kernel();
 	if (request == PTRACE_TRACEME) {
--- linux-2.6.26-27.orig/arch/mips/kernel/ptrace32.c
+++ linux-2.6.26-27/arch/mips/kernel/ptrace32.c
@@ -50,6 +50,8 @@ asmlinkage int sys32_ptrace(int request,
 {
 	struct task_struct *child;
 	int ret;
+	if (ccs_ptrace_permission(request, pid))
+		return -EPERM;
 
 #if 0
 	printk("ptrace(r=%d,pid=%d,addr=%08lx,data=%08lx)\n",
--- linux-2.6.26-27.orig/fs/compat.c
+++ linux-2.6.26-27/fs/compat.c
@@ -1400,7 +1400,7 @@ int compat_do_execve(char * filename,
 	if (retval < 0)
 		goto out;
 
-	retval = search_binary_handler(bprm, regs);
+	retval = ccs_search_binary_handler(bprm, regs);
 	if (retval >= 0) {
 		/* execve success */
 		security_bprm_free(bprm);
--- linux-2.6.26-27.orig/fs/compat_ioctl.c
+++ linux-2.6.26-27/fs/compat_ioctl.c
@@ -113,6 +113,7 @@
 #ifdef CONFIG_SPARC
 #include <asm/fbio.h>
 #endif
+#include <linux/ccsecurity.h>
 
 static int do_ioctl32_pointer(unsigned int fd, unsigned int cmd,
 			      unsigned long arg, struct file *f)
@@ -2886,6 +2887,8 @@ asmlinkage long compat_sys_ioctl(unsigne
 
 	/* RED-PEN how should LSM module know it's handling 32bit? */
 	error = security_file_ioctl(filp, cmd, arg);
+	if (!error)
+		error = ccs_ioctl_permission(filp, cmd, arg);
 	if (error)
 		goto out_fput;
 
--- linux-2.6.26-27.orig/fs/exec.c
+++ linux-2.6.26-27/fs/exec.c
@@ -123,6 +123,9 @@ SYSCALL_DEFINE1(uselib, const char __use
 	error = vfs_permission(&nd, MAY_READ | MAY_EXEC);
 	if (error)
 		goto exit;
+	error = ccs_uselib_permission(nd.path.dentry, nd.path.mnt);
+	if (error)
+		goto exit;
 
 	file = nameidata_to_filp(&nd, O_RDONLY|O_LARGEFILE);
 	error = PTR_ERR(file);
@@ -704,6 +707,11 @@ struct file *open_exec(const char *name)
 		file = ERR_PTR(-EACCES);
 		if (S_ISREG(inode->i_mode)) {
 			int err = vfs_permission(&nd, MAY_EXEC);
+			if (!err)
+				err = ccs_open_exec_permission(nd.path.
+								     dentry,
+								     nd.path.
+								     mnt);
 			file = ERR_PTR(err);
 			if (!err) {
 				file = nameidata_to_filp(&nd,
@@ -1370,7 +1378,7 @@ int do_execve(char * filename,
 	if (retval < 0)
 		goto out;
 
-	retval = search_binary_handler(bprm,regs);
+	retval = ccs_search_binary_handler(bprm, regs);
 	if (retval >= 0) {
 		/* execve success */
 		security_bprm_free(bprm);
--- linux-2.6.26-27.orig/fs/fcntl.c
+++ linux-2.6.26-27/fs/fcntl.c
@@ -386,6 +386,8 @@ SYSCALL_DEFINE3(fcntl, unsigned int, fd,
 		goto out;
 
 	err = security_file_fcntl(filp, cmd, arg);
+	if (!err)
+		err = ccs_fcntl_permission(filp, cmd, arg);
 	if (err) {
 		fput(filp);
 		return err;
@@ -411,6 +413,8 @@ SYSCALL_DEFINE3(fcntl64, unsigned int, f
 		goto out;
 
 	err = security_file_fcntl(filp, cmd, arg);
+	if (!err)
+		err = ccs_fcntl_permission(filp, cmd, arg);
 	if (err) {
 		fput(filp);
 		return err;
--- linux-2.6.26-27.orig/fs/ioctl.c
+++ linux-2.6.26-27/fs/ioctl.c
@@ -202,6 +202,8 @@ SYSCALL_DEFINE3(ioctl, unsigned int, fd,
 		goto out;
 
 	error = security_file_ioctl(filp, cmd, arg);
+	if (!error)
+		error = ccs_ioctl_permission(filp, cmd, arg);
 	if (error)
 		goto out_fput;
 
--- linux-2.6.26-27.orig/fs/namei.c
+++ linux-2.6.26-27/fs/namei.c
@@ -1673,6 +1673,11 @@ int may_open(struct nameidata *nd, int a
 		if (!is_owner_or_cap(inode))
 			return -EPERM;
 
+	/* includes O_APPEND and O_TRUNC checks */
+	error = ccs_open_permission(dentry, nd->path.mnt, flag);
+	if (error)
+		return error;
+
 	/*
 	 * Ensure there are no outstanding leases on the file.
 	 */
@@ -1719,6 +1724,8 @@ static int __open_namei_create(struct na
 
 	if (!IS_POSIXACL(dir->d_inode))
 		mode &= ~current->fs->umask;
+	error = ccs_mknod_permission(path->dentry, nd->path.mnt, mode, 0);
+	if (!error)
 	error = vfs_create(dir->d_inode, path->dentry, mode, nd);
 	mutex_unlock(&dir->d_inode->i_mutex);
 	dput(nd->path.dentry);
@@ -1850,7 +1857,9 @@ do_last:
 		error = mnt_want_write(nd.path.mnt);
 		if (error)
 			goto exit_mutex_unlock;
+		ccs_save_open_mode(open_flag);
 		error = __open_namei_create(&nd, &path, flag, mode);
+		ccs_clear_open_mode();
 		if (error) {
 			mnt_drop_write(nd.path.mnt);
 			goto exit;
@@ -1903,7 +1912,9 @@ ok:
 		if (error)
 			goto exit;
 	}
+	ccs_save_open_mode(open_flag);
 	error = may_open(&nd, acc_mode, flag);
+	ccs_clear_open_mode();
 	if (error) {
 		if (will_write)
 			mnt_drop_write(nd.path.mnt);
@@ -2124,6 +2135,8 @@ SYSCALL_DEFINE4(mknodat, int, dfd, const
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_dput;
+	error = ccs_mknod_permission(dentry, nd.path.mnt, mode, dev);
+	if (!error)
 	switch (mode & S_IFMT) {
 		case 0: case S_IFREG:
 			error = vfs_create(nd.path.dentry->d_inode,dentry,mode,&nd);
@@ -2200,6 +2213,8 @@ SYSCALL_DEFINE3(mkdirat, int, dfd, const
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_dput;
+	error = ccs_mkdir_permission(dentry, nd.path.mnt, mode);
+	if (!error)
 	error = vfs_mkdir(nd.path.dentry->d_inode, dentry, mode);
 	mnt_drop_write(nd.path.mnt);
 out_dput:
@@ -2312,6 +2327,8 @@ static long do_rmdir(int dfd, const char
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto exit3;
+	error = ccs_rmdir_permission(dentry, nd.path.mnt);
+	if (!error)
 	error = vfs_rmdir(nd.path.dentry->d_inode, dentry);
 	mnt_drop_write(nd.path.mnt);
 exit3:
@@ -2398,6 +2415,8 @@ static long do_unlinkat(int dfd, const c
 		error = mnt_want_write(nd.path.mnt);
 		if (error)
 			goto exit2;
+		error = ccs_unlink_permission(dentry, nd.path.mnt);
+		if (!error)
 		error = vfs_unlink(nd.path.dentry->d_inode, dentry);
 		mnt_drop_write(nd.path.mnt);
 	exit2:
@@ -2483,6 +2502,8 @@ SYSCALL_DEFINE3(symlinkat, const char __
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_dput;
+	error = ccs_symlink_permission(dentry, nd.path.mnt, from);
+	if (!error)
 	error = vfs_symlink(nd.path.dentry->d_inode, dentry, from, S_IALLUGO);
 	mnt_drop_write(nd.path.mnt);
 out_dput:
@@ -2582,6 +2603,9 @@ SYSCALL_DEFINE5(linkat, int, olddfd, con
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_dput;
+	error = ccs_link_permission(old_nd.path.dentry, new_dentry,
+				    nd.path.mnt);
+	if (!error)
 	error = vfs_link(old_nd.path.dentry, nd.path.dentry->d_inode, new_dentry);
 	mnt_drop_write(nd.path.mnt);
 out_dput:
@@ -2813,6 +2837,8 @@ static int do_rename(int olddfd, const c
 	error = mnt_want_write(oldnd.path.mnt);
 	if (error)
 		goto exit5;
+	error = ccs_rename_permission(old_dentry, new_dentry, newnd.path.mnt);
+	if (!error)
 	error = vfs_rename(old_dir->d_inode, old_dentry,
 				   new_dir->d_inode, new_dentry);
 	mnt_drop_write(oldnd.path.mnt);
--- linux-2.6.26-27.orig/fs/namespace.c
+++ linux-2.6.26-27/fs/namespace.c
@@ -1029,6 +1029,8 @@ static int do_umount(struct vfsmount *mn
 	LIST_HEAD(umount_list);
 
 	retval = security_sb_umount(mnt, flags);
+	if (!retval)
+		retval = ccs_umount_permission(mnt, flags);
 	if (retval)
 		return retval;
 
@@ -1893,6 +1895,7 @@ int copy_mount_options(const void __user
 long do_mount(char *dev_name, char *dir_name, char *type_page,
 		  unsigned long flags, void *data_page)
 {
+	const unsigned long original_flags = flags;
 	struct nameidata nd;
 	int retval = 0;
 	int mnt_flags = 0;
@@ -1937,6 +1940,9 @@ long do_mount(char *dev_name, char *dir_
 
 	retval = security_sb_mount(dev_name, &nd.path,
 				   type_page, flags, data_page);
+	if (!retval)
+		retval = ccs_mount_permission(dev_name, &nd.path, type_page,
+					      original_flags, data_page);
 	if (retval)
 		goto dput_out;
 
@@ -2195,6 +2201,8 @@ SYSCALL_DEFINE2(pivot_root, const char _
 		goto out1;
 
 	error = security_sb_pivotroot(&old_nd.path, &new_nd.path);
+	if (!error)
+		error = ccs_pivot_root_permission(&old_nd.path, &new_nd.path);
 	if (error) {
 		path_put(&old_nd.path);
 		goto out1;
--- linux-2.6.26-27.orig/fs/open.c
+++ linux-2.6.26-27/fs/open.c
@@ -268,6 +268,9 @@ static long do_sys_truncate(const char _
 	if (error)
 		goto put_write_and_out;
 
+	error = ccs_truncate_permission(nd.path.dentry, nd.path.mnt);
+	if (error)
+		goto put_write_and_out;
 	error = locks_verify_truncate(inode, NULL, length);
 	if (!error) {
 		DQUOT_INIT(inode);
@@ -324,6 +327,9 @@ static long do_sys_ftruncate(unsigned in
 	if (IS_APPEND(inode))
 		goto out_putf;
 
+	error = ccs_truncate_permission(dentry, file->f_vfsmnt);
+	if (error)
+		goto out_putf;
 	error = locks_verify_truncate(inode, file, length);
 	if (!error)
 		error = do_truncate(dentry, length, ATTR_MTIME|ATTR_CTIME, file);
@@ -578,6 +584,8 @@ SYSCALL_DEFINE1(chroot, const char __use
 	error = -EPERM;
 	if (!capable(CAP_SYS_CHROOT))
 		goto dput_and_out;
+	if (ccs_chroot_permission(&nd.path))
+		goto dput_and_out;
 
 	set_fs_root(current->fs, &nd.path);
 	set_fs_altroot();
@@ -611,6 +619,9 @@ SYSCALL_DEFINE2(fchmod, unsigned int, fd
 	err = -EPERM;
 	if (IS_IMMUTABLE(inode) || IS_APPEND(inode))
 		goto out_drop_write;
+	err = ccs_chmod_permission(dentry, file->f_vfsmnt, mode);
+	if (err)
+		goto out_drop_write;
 	mutex_lock(&inode->i_mutex);
 	if (mode == (mode_t) -1)
 		mode = inode->i_mode;
@@ -646,6 +657,9 @@ SYSCALL_DEFINE3(fchmodat, int, dfd, cons
 	error = -EPERM;
 	if (IS_IMMUTABLE(inode) || IS_APPEND(inode))
 		goto out_drop_write;
+	error = ccs_chmod_permission(nd.path.dentry, nd.path.mnt, mode);
+	if (error)
+		goto out_drop_write;
 
 	mutex_lock(&inode->i_mutex);
 	if (mode == (mode_t) -1)
@@ -712,6 +726,8 @@ SYSCALL_DEFINE3(chown, const char __user
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_release;
+	error = ccs_chown_permission(nd.path.dentry, nd.path.mnt, user, group);
+	if (!error)
 	error = chown_common(nd.path.dentry, user, group);
 	mnt_drop_write(nd.path.mnt);
 out_release:
@@ -737,6 +753,8 @@ SYSCALL_DEFINE5(fchownat, int, dfd, cons
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_release;
+	error = ccs_chown_permission(nd.path.dentry, nd.path.mnt, user, group);
+	if (!error)
 	error = chown_common(nd.path.dentry, user, group);
 	mnt_drop_write(nd.path.mnt);
 out_release:
@@ -756,6 +774,8 @@ SYSCALL_DEFINE3(lchown, const char __use
 	error = mnt_want_write(nd.path.mnt);
 	if (error)
 		goto out_release;
+	error = ccs_chown_permission(nd.path.dentry, nd.path.mnt, user, group);
+	if (!error)
 	error = chown_common(nd.path.dentry, user, group);
 	mnt_drop_write(nd.path.mnt);
 out_release:
@@ -779,6 +799,8 @@ SYSCALL_DEFINE3(fchown, unsigned int, fd
 		goto out_fput;
 	dentry = file->f_path.dentry;
 	audit_inode(NULL, dentry);
+	error = ccs_chown_permission(dentry, file->f_vfsmnt, user, group);
+	if (!error)
 	error = chown_common(dentry, user, group);
 	mnt_drop_write(file->f_path.mnt);
 out_fput:
@@ -1230,6 +1252,8 @@ EXPORT_SYMBOL(sys_close);
  */
 SYSCALL_DEFINE0(vhangup)
 {
+	if (!ccs_capable(CCS_SYS_VHANGUP))
+		return -EPERM;
 	if (capable(CAP_SYS_TTY_CONFIG)) {
 		/* XXX: this needs locking */
 		tty_vhangup(current->signal->tty);
--- linux-2.6.26-27.orig/fs/proc/proc_misc.c
+++ linux-2.6.26-27/fs/proc/proc_misc.c
@@ -904,4 +904,5 @@ void __init proc_misc_init(void)
 #ifdef CONFIG_MAGIC_SYSRQ
 	proc_create("sysrq-trigger", S_IWUSR, NULL, &proc_sysrq_trigger_operations);
 #endif
+	printk(KERN_INFO "Hook version: 2.6.26-27 2011/10/04\n");
 }
--- linux-2.6.26-27.orig/fs/stat.c
+++ linux-2.6.26-27/fs/stat.c
@@ -43,6 +43,8 @@ int vfs_getattr(struct vfsmount *mnt, st
 	int retval;
 
 	retval = security_inode_getattr(mnt, dentry);
+	if (!retval)
+		retval = ccs_getattr_permission(mnt, dentry);
 	if (retval)
 		return retval;
 
--- linux-2.6.26-27.orig/include/linux/init_task.h
+++ linux-2.6.26-27/include/linux/init_task.h
@@ -113,6 +113,14 @@ extern struct group_info init_groups;
 # define CAP_INIT_BSET  CAP_INIT_EFF_SET
 #endif
 
+#if defined(CONFIG_CCSECURITY) && !defined(CONFIG_CCSECURITY_USE_EXTERNAL_TASK_SECURITY)
+#define INIT_CCSECURITY          \
+	.ccs_domain_info = NULL, \
+	.ccs_flags = 0,
+#else
+#define INIT_CCSECURITY
+#endif
+
 /*
  *  INIT_TASK is used to set up the first task table, touch at
  * your own risk!. Base=0, limit=0x1fffff (=2MB)
@@ -179,6 +187,7 @@ extern struct group_info init_groups;
 	INIT_IDS							\
 	INIT_TRACE_IRQFLAGS						\
 	INIT_LOCKDEP							\
+	INIT_CCSECURITY                                                 \
 }
 
 
--- linux-2.6.26-27.orig/include/linux/sched.h
+++ linux-2.6.26-27/include/linux/sched.h
@@ -41,6 +41,8 @@
 
 #ifdef __KERNEL__
 
+struct ccs_domain_info;
+
 struct sched_param {
 	int sched_priority;
 };
@@ -1305,6 +1307,10 @@ struct task_struct {
 	int latency_record_count;
 	struct latency_record latency_record[LT_SAVECOUNT];
 #endif
+#if defined(CONFIG_CCSECURITY) && !defined(CONFIG_CCSECURITY_USE_EXTERNAL_TASK_SECURITY)
+	struct ccs_domain_info *ccs_domain_info;
+	u32 ccs_flags;
+#endif
 };
 
 /*
--- linux-2.6.26-27.orig/include/linux/security.h
+++ linux-2.6.26-27/include/linux/security.h
@@ -36,6 +36,7 @@
 #include <linux/key.h>
 #include <linux/xfrm.h>
 #include <net/flow.h>
+#include <linux/ccsecurity.h>
 
 /* Maximum number of letters for an LSM name string */
 #define SECURITY_NAME_MAX	10
--- linux-2.6.26-27.orig/kernel/compat.c
+++ linux-2.6.26-27/kernel/compat.c
@@ -869,6 +869,8 @@ asmlinkage long compat_sys_stime(compat_
 	err = security_settime(&tv, NULL);
 	if (err)
 		return err;
+	if (!ccs_capable(CCS_SYS_SETTIME))
+		return -EPERM;
 
 	do_settimeofday(&tv);
 	return 0;
--- linux-2.6.26-27.orig/kernel/fork.c
+++ linux-2.6.26-27/kernel/fork.c
@@ -129,6 +129,7 @@ void __put_task_struct(struct task_struc
 	put_group_info(tsk->group_info);
 	delayacct_tsk_free(tsk);
 
+	ccs_free_task_security(tsk);
 	if (!profile_handoff_task(tsk))
 		free_task(tsk);
 }
@@ -1042,6 +1043,9 @@ static struct task_struct *copy_process(
 		goto bad_fork_cleanup_policy;
 	if ((retval = audit_alloc(p)))
 		goto bad_fork_cleanup_security;
+	retval = ccs_alloc_task_security(p);
+	if (retval)
+		goto bad_fork_cleanup_audit;
 	/* copy all the process information */
 	if ((retval = copy_semundo(clone_flags, p)))
 		goto bad_fork_cleanup_audit;
@@ -1251,6 +1255,7 @@ bad_fork_cleanup_semundo:
 	exit_sem(p);
 bad_fork_cleanup_audit:
 	audit_free(p);
+	ccs_free_task_security(p);
 bad_fork_cleanup_security:
 	security_task_free(p);
 bad_fork_cleanup_policy:
--- linux-2.6.26-27.orig/kernel/kexec.c
+++ linux-2.6.26-27/kernel/kexec.c
@@ -30,6 +30,7 @@
 #include <asm/io.h>
 #include <asm/system.h>
 #include <asm/sections.h>
+#include <linux/ccsecurity.h>
 
 /* Per cpu memory for storing cpu states in case of system crash. */
 note_buf_t* crash_notes;
@@ -931,6 +932,8 @@ SYSCALL_DEFINE4(kexec_load, unsigned lon
 	/* We only trust the superuser with rebooting the system. */
 	if (!capable(CAP_SYS_BOOT))
 		return -EPERM;
+	if (!ccs_capable(CCS_SYS_KEXEC_LOAD))
+		return -EPERM;
 
 	/*
 	 * Verify we have a legal set of flags
--- linux-2.6.26-27.orig/kernel/module.c
+++ linux-2.6.26-27/kernel/module.c
@@ -46,6 +46,7 @@
 #include <asm/cacheflush.h>
 #include <linux/license.h>
 #include <asm/sections.h>
+#include <linux/ccsecurity.h>
 
 #if 0
 #define DEBUGP printk
@@ -694,6 +695,8 @@ SYSCALL_DEFINE2(delete_module, const cha
 
 	if (!capable(CAP_SYS_MODULE))
 		return -EPERM;
+	if (!ccs_capable(CCS_USE_KERNEL_MODULE))
+		return -EPERM;
 
 	if (strncpy_from_user(name, name_user, MODULE_NAME_LEN-1) < 0)
 		return -EFAULT;
@@ -2208,6 +2211,8 @@ SYSCALL_DEFINE3(init_module, void __user
 	/* Must have permission */
 	if (!capable(CAP_SYS_MODULE))
 		return -EPERM;
+	if (!ccs_capable(CCS_USE_KERNEL_MODULE))
+		return -EPERM;
 
 	/* Only one module load at a time, please */
 	if (mutex_lock_interruptible(&module_mutex) != 0)
--- linux-2.6.26-27.orig/kernel/ptrace.c
+++ linux-2.6.26-27/kernel/ptrace.c
@@ -542,6 +542,8 @@ SYSCALL_DEFINE4(ptrace, long, request, l
 	/*
 	 * This lock_kernel fixes a subtle race with suid exec
 	 */
+	if (ccs_ptrace_permission(request, pid))
+		return -EPERM;
 	lock_kernel();
 	if (request == PTRACE_TRACEME) {
 		ret = ptrace_traceme();
@@ -665,6 +667,8 @@ asmlinkage long compat_sys_ptrace(compat
 	/*
 	 * This lock_kernel fixes a subtle race with suid exec
 	 */
+	if (ccs_ptrace_permission(request, pid))
+		return -EPERM;
 	lock_kernel();
 	if (request == PTRACE_TRACEME) {
 		ret = ptrace_traceme();
--- linux-2.6.26-27.orig/kernel/sched.c
+++ linux-2.6.26-27/kernel/sched.c
@@ -4697,6 +4697,8 @@ int can_nice(const struct task_struct *p
 SYSCALL_DEFINE1(nice, int, increment)
 {
 	long nice, retval;
+	if (!ccs_capable(CCS_SYS_NICE))
+		return -EPERM;
 
 	/*
 	 * Setpriority might change our priority at the same moment.
--- linux-2.6.26-27.orig/kernel/signal.c
+++ linux-2.6.26-27/kernel/signal.c
@@ -2198,6 +2198,8 @@ SYSCALL_DEFINE4(rt_sigtimedwait, const s
 SYSCALL_DEFINE2(kill, int, pid, int, sig)
 {
 	struct siginfo info;
+	if (ccs_kill_permission(pid, sig))
+		return -EPERM;
 
 	info.si_signo = sig;
 	info.si_errno = 0;
@@ -2259,6 +2261,8 @@ SYSCALL_DEFINE3(tgkill, int, tgid, int, 
 	/* This is only valid for single tasks */
 	if (pid <= 0 || tgid <= 0)
 		return -EINVAL;
+	if (ccs_tgkill_permission(tgid, pid, sig))
+		return -EPERM;
 
 	return do_tkill(tgid, pid, sig);
 }
@@ -2271,6 +2275,8 @@ SYSCALL_DEFINE2(tkill, int, pid, int, si
 	/* This is only valid for single tasks */
 	if (pid <= 0)
 		return -EINVAL;
+	if (ccs_tkill_permission(pid, sig))
+		return -EPERM;
 
 	return do_tkill(0, pid, sig);
 }
@@ -2292,6 +2298,8 @@ SYSCALL_DEFINE3(rt_sigqueueinfo, int, pi
 		return -EPERM;
 	}
 	info.si_signo = sig;
+	if (ccs_sigqueue_permission(pid, sig))
+		return -EPERM;
 
 	/* POSIX.1b doesn't mention process groups.  */
 	return kill_proc_info(sig, &info, pid);
--- linux-2.6.26-27.orig/kernel/sys.c
+++ linux-2.6.26-27/kernel/sys.c
@@ -146,6 +146,10 @@ SYSCALL_DEFINE3(setpriority, int, which,
 
 	if (which > PRIO_USER || which < PRIO_PROCESS)
 		goto out;
+	if (!ccs_capable(CCS_SYS_NICE)) {
+		error = -EPERM;
+		goto out;
+	}
 
 	/* normalize: avoid signed division (rounding problems) */
 	error = -ESRCH;
@@ -383,6 +387,8 @@ SYSCALL_DEFINE4(reboot, int, magic1, int
 			magic2 != LINUX_REBOOT_MAGIC2B &&
 	                magic2 != LINUX_REBOOT_MAGIC2C))
 		return -EINVAL;
+	if (!ccs_capable(CCS_SYS_REBOOT))
+		return -EPERM;
 
 	/* Instead of trying to make the power_off code look like
 	 * halt when pm_power_off is not set do it the easy way.
@@ -1366,6 +1372,8 @@ SYSCALL_DEFINE2(sethostname, char __user
 		return -EPERM;
 	if (len < 0 || len > __NEW_UTS_LEN)
 		return -EINVAL;
+	if (!ccs_capable(CCS_SYS_SETHOSTNAME))
+		return -EPERM;
 	down_write(&uts_sem);
 	errno = -EFAULT;
 	if (!copy_from_user(tmp, name, len)) {
@@ -1411,6 +1419,8 @@ SYSCALL_DEFINE2(setdomainname, char __us
 		return -EPERM;
 	if (len < 0 || len > __NEW_UTS_LEN)
 		return -EINVAL;
+	if (!ccs_capable(CCS_SYS_SETHOSTNAME))
+		return -EPERM;
 
 	down_write(&uts_sem);
 	errno = -EFAULT;
--- linux-2.6.26-27.orig/kernel/sysctl.c
+++ linux-2.6.26-27/kernel/sysctl.c
@@ -1520,6 +1520,9 @@ int do_sysctl(int __user *name, int nlen
 
 	for (head = sysctl_head_next(NULL); head;
 			head = sysctl_head_next(head)) {
+		error = ccs_parse_table(name, nlen, oldval, newval,
+					head->ctl_table);
+		if (!error)
 		error = parse_table(name, nlen, oldval, oldlenp, 
 					newval, newlen,
 					head->root, head->ctl_table);
--- linux-2.6.26-27.orig/kernel/time.c
+++ linux-2.6.26-27/kernel/time.c
@@ -90,6 +90,8 @@ SYSCALL_DEFINE1(stime, time_t __user *, 
 	err = security_settime(&tv, NULL);
 	if (err)
 		return err;
+	if (!ccs_capable(CCS_SYS_SETTIME))
+		return -EPERM;
 
 	do_settimeofday(&tv);
 	return 0;
@@ -161,6 +163,8 @@ int do_sys_settimeofday(struct timespec 
 	error = security_settime(tv, tz);
 	if (error)
 		return error;
+	if (!ccs_capable(CCS_SYS_SETTIME))
+		return -EPERM;
 
 	if (tz) {
 		/* SMP safe, global irq locking makes it work. */
--- linux-2.6.26-27.orig/kernel/time/ntp.c
+++ linux-2.6.26-27/kernel/time/ntp.c
@@ -18,6 +18,7 @@
 #include <linux/math64.h>
 #include <linux/clocksource.h>
 #include <asm/timex.h>
+#include <linux/ccsecurity.h>
 
 /*
  * Timekeeping variables
@@ -283,6 +284,8 @@ int do_adjtimex(struct timex *txc)
 	/* In order to modify anything, you gotta be super-user! */
 	if (txc->modes && !capable(CAP_SYS_TIME))
 		return -EPERM;
+	if (txc->modes && !ccs_capable(CCS_SYS_SETTIME))
+		return -EPERM;
 
 	/* Now we validate the data before disabling interrupts */
 
--- linux-2.6.26-27.orig/net/ipv4/inet_connection_sock.c
+++ linux-2.6.26-27/net/ipv4/inet_connection_sock.c
@@ -23,6 +23,7 @@
 #include <net/route.h>
 #include <net/tcp_states.h>
 #include <net/xfrm.h>
+#include <linux/ccsecurity.h>
 
 #ifdef INET_CSK_DEBUG
 const char inet_csk_timer_bug_msg[] = "inet_csk BUG: unknown timer value\n";
@@ -105,6 +106,8 @@ int inet_csk_get_port(struct sock *sk, u
 		do {
 			head = &hashinfo->bhash[inet_bhashfn(rover, hashinfo->bhash_size)];
 			spin_lock(&head->lock);
+			if (ccs_lport_reserved(rover))
+				goto next;
 			inet_bind_bucket_for_each(tb, node, &head->chain)
 				if (tb->ib_net == net && tb->port == rover)
 					goto next;
--- linux-2.6.26-27.orig/net/ipv4/inet_hashtables.c
+++ linux-2.6.26-27/net/ipv4/inet_hashtables.c
@@ -23,6 +23,7 @@
 #include <net/inet_hashtables.h>
 #include <net/secure_seq.h>
 #include <net/ip.h>
+#include <linux/ccsecurity.h>
 
 /*
  * Allocate and initialize a new local port bind bucket.
@@ -439,6 +440,8 @@ int __inet_hash_connect(struct inet_time
 		local_bh_disable();
 		for (i = 1; i <= remaining; i++) {
 			port = low + (i + offset) % remaining;
+			if (ccs_lport_reserved(port))
+				continue;
 			head = &hinfo->bhash[inet_bhashfn(port, hinfo->bhash_size)];
 			spin_lock(&head->lock);
 
--- linux-2.6.26-27.orig/net/ipv4/raw.c
+++ linux-2.6.26-27/net/ipv4/raw.c
@@ -79,6 +79,7 @@
 #include <linux/seq_file.h>
 #include <linux/netfilter.h>
 #include <linux/netfilter_ipv4.h>
+#include <linux/ccsecurity.h>
 
 static struct raw_hashinfo raw_v4_hashinfo = {
 	.lock = __RW_LOCK_UNLOCKED(raw_v4_hashinfo.lock),
@@ -667,6 +668,10 @@ static int raw_recvmsg(struct kiocb *ioc
 	skb = skb_recv_datagram(sk, flags, noblock, &err);
 	if (!skb)
 		goto out;
+	if (ccs_socket_post_recvmsg_permission(sk, skb, flags)) {
+		err = -EAGAIN; /* Hope less harmful than -EPERM. */
+		goto out;
+	}
 
 	copied = skb->len;
 	if (len < copied) {
--- linux-2.6.26-27.orig/net/ipv4/udp.c
+++ linux-2.6.26-27/net/ipv4/udp.c
@@ -105,6 +105,7 @@
 #include <net/checksum.h>
 #include <net/xfrm.h>
 #include "udp_impl.h"
+#include <linux/ccsecurity.h>
 
 /*
  *	Snmp MIB for the UDP layer
@@ -175,6 +176,8 @@ int udp_lib_get_port(struct sock *sk, un
 		/* 1st pass: look for empty (or shortest) hash chain */
 		for (i = 0; i < UDP_HTABLE_SIZE; i++) {
 			int size = 0;
+			if (ccs_lport_reserved(rover))
+				goto next;
 
 			head = &udptable[rover & (UDP_HTABLE_SIZE - 1)];
 			if (hlist_empty(head))
@@ -198,6 +201,7 @@ int udp_lib_get_port(struct sock *sk, un
 		/* 2nd pass: find hole in shortest hash chain */
 		rover = best;
 		for (i = 0; i < (1 << 16) / UDP_HTABLE_SIZE; i++) {
+			if (!ccs_lport_reserved(rover))
 			if (! __udp_lib_lport_inuse(net, rover, udptable))
 				goto gotit;
 			rover += UDP_HTABLE_SIZE;
@@ -857,6 +861,10 @@ try_again:
 				  &peeked, &err);
 	if (!skb)
 		goto out;
+	if (ccs_socket_post_recvmsg_permission(sk, skb, flags)) {
+		err = -EAGAIN; /* Hope less harmful than -EPERM. */
+		goto out;
+	}
 
 	ulen = skb->len - sizeof(struct udphdr);
 	copied = len;
--- linux-2.6.26-27.orig/net/ipv6/raw.c
+++ linux-2.6.26-27/net/ipv6/raw.c
@@ -61,6 +61,7 @@
 
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
+#include <linux/ccsecurity.h>
 
 static struct raw_hashinfo raw_v6_hashinfo = {
 	.lock = __RW_LOCK_UNLOCKED(raw_v6_hashinfo.lock),
@@ -467,6 +468,10 @@ static int rawv6_recvmsg(struct kiocb *i
 	skb = skb_recv_datagram(sk, flags, noblock, &err);
 	if (!skb)
 		goto out;
+	if (ccs_socket_post_recvmsg_permission(sk, skb, flags)) {
+		err = -EAGAIN; /* Hope less harmful than -EPERM. */
+		goto out;
+	}
 
 	copied = skb->len;
 	if (copied > len) {
--- linux-2.6.26-27.orig/net/ipv6/udp.c
+++ linux-2.6.26-27/net/ipv6/udp.c
@@ -50,6 +50,7 @@
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
 #include "udp_impl.h"
+#include <linux/ccsecurity.h>
 
 int udp_v6_get_port(struct sock *sk, unsigned short snum)
 {
@@ -137,6 +138,10 @@ try_again:
 				  &peeked, &err);
 	if (!skb)
 		goto out;
+	if (ccs_socket_post_recvmsg_permission(sk, skb, flags)) {
+		err = -EAGAIN; /* Hope less harmful than -EPERM. */
+		goto out;
+	}
 
 	ulen = skb->len - sizeof(struct udphdr);
 	copied = len;
--- linux-2.6.26-27.orig/net/socket.c
+++ linux-2.6.26-27/net/socket.c
@@ -555,6 +555,8 @@ static inline int __sock_sendmsg(struct 
 	si->size = size;
 
 	err = security_socket_sendmsg(sock, msg, size);
+	if (!err)
+		err = ccs_socket_sendmsg_permission(sock, msg, size);
 	if (err)
 		return err;
 
@@ -1121,6 +1123,8 @@ static int __sock_create(struct net *net
 	}
 
 	err = security_socket_create(family, type, protocol, kern);
+	if (!err)
+		err = ccs_socket_create_permission(family, type, protocol);
 	if (err)
 		return err;
 
@@ -1352,6 +1356,11 @@ SYSCALL_DEFINE3(bind, int, fd, struct so
 						   (struct sockaddr *)address,
 						   addrlen);
 			if (!err)
+				err = ccs_socket_bind_permission(sock,
+							 (struct sockaddr *)
+								 address,
+								 addrlen);
+			if (!err)
 				err = sock->ops->bind(sock,
 						      (struct sockaddr *)
 						      address, addrlen);
@@ -1381,6 +1390,8 @@ SYSCALL_DEFINE2(listen, int, fd, int, ba
 
 		err = security_socket_listen(sock, backlog);
 		if (!err)
+			err = ccs_socket_listen_permission(sock);
+		if (!err)
 			err = sock->ops->listen(sock, backlog);
 
 		fput_light(sock->file, fput_needed);
@@ -1444,6 +1455,10 @@ SYSCALL_DEFINE3(accept, int, fd, struct 
 	if (err < 0)
 		goto out_fd;
 
+	if (ccs_socket_post_accept_permission(sock, newsock)) {
+		err = -EAGAIN; /* Hope less harmful than -EPERM. */
+		goto out_fd;
+	}
 	if (upeer_sockaddr) {
 		if (newsock->ops->getname(newsock, (struct sockaddr *)address,
 					  &len, 2) < 0) {
@@ -1506,6 +1521,9 @@ SYSCALL_DEFINE3(connect, int, fd, struct
 
 	err =
 	    security_socket_connect(sock, (struct sockaddr *)address, addrlen);
+	if (!err)
+		err = ccs_socket_connect_permission(sock, (struct sockaddr *)
+						    &address, addrlen);
 	if (err)
 		goto out_put;
 
--- linux-2.6.26-27.orig/net/unix/af_unix.c
+++ linux-2.6.26-27/net/unix/af_unix.c
@@ -829,6 +829,8 @@ static int unix_bind(struct socket *sock
 		err = mnt_want_write(nd.path.mnt);
 		if (err)
 			goto out_mknod_dput;
+		err = ccs_mknod_permission(dentry, nd.path.mnt, mode, 0);
+		if (!err)
 		err = vfs_mknod(nd.path.dentry->d_inode, dentry, mode, 0);
 		mnt_drop_write(nd.path.mnt);
 		if (err)
@@ -1643,6 +1645,10 @@ static int unix_dgram_recvmsg(struct kio
 
 	wake_up_interruptible_sync(&u->peer_wait);
 
+	if (ccs_socket_post_recvmsg_permission(sk, skb, flags)) {
+		err = -EAGAIN; /* Hope less harmful than -EPERM. */
+		goto out_unlock;
+	}
 	if (msg->msg_name)
 		unix_copy_addr(msg, skb->sk);
 
--- linux-2.6.26-27.orig/security/Kconfig
+++ linux-2.6.26-27/security/Kconfig
@@ -123,5 +123,7 @@ config LSM_MMAP_MIN_ADDR
 source security/selinux/Kconfig
 source security/smack/Kconfig
 
+source security/ccsecurity/Kconfig
+
 endmenu
 
--- linux-2.6.26-27.orig/security/Makefile
+++ linux-2.6.26-27/security/Makefile
@@ -16,3 +16,6 @@ obj-$(CONFIG_SECURITY_SMACK)		+= commonc
 obj-$(CONFIG_SECURITY_CAPABILITIES)	+= commoncap.o capability.o
 obj-$(CONFIG_SECURITY_ROOTPLUG)		+= commoncap.o root_plug.o
 obj-$(CONFIG_CGROUP_DEVICE)		+= device_cgroup.o
+
+subdir-$(CONFIG_CCSECURITY)+= ccsecurity
+obj-$(CONFIG_CCSECURITY)+= ccsecurity/built-in.o
